### PR TITLE
Parser update

### LINKED ISSUE #521: Bad error messages
These are found by @fionakillalea - thanks!

> Two areas that resulted in the "This should never happen, sorry!" errors in the program. 
>  
> The first occurs with the flatten function when the arguments to the function are in the wrong order (the int should be the first argument) as demonstrated in an example below. 
>  
> find b : bool 
> such that b = ([1, 2, 3] = flatten([1, 2, 3], 0))
> 
> 
> The second bug I found occurs with the alldifferent_except function when the second operator is missing, as shown below. 
>  
> find c : bool 
> such that c = alldifferent_except([1, 2, 5, 1, 6]) 
>

### LINKED ISSUE #554: Remove `conjure tsdef`
Once #546 is merged.

### PR DESCRIPTION
**Pull request by Maurice-Byrne**
The big Conjure parser overhaul.
Not quite ready for merging yet but tentatively putting it up for review.
The LSP function is under a separate PR.

Changes:
- The parser has been updated to be error tolerant, at a minimum finding the first ambiguous syntax error in each top level statement, and often much better than that.
- The parsing pipeline has been adjusted to retain more information for longer so that error messages can be more specific.
- Notably, type checking is now performed with access to location and structural information meaning no more hard to interpret de-sugared type errors.
  - The old type checker is still run as a fall-back so there is no loss in thoroughness. 
- Groundwork set for LSP support, implemented in #547.
- Pretty printer now respects comments and structural choices as it is run on the parse tree not the generated model.

Known problems:
- Pretty printer is not very pretty at the moment.
- Type checking currently does not work for solution files that rely on parameters.
  - This does not impact functionality as these files are processed without type checking validation however. 
- TypeScript output is broken.

Testing:
- This update breaks a lot of tests, especially those that use error messages. 
- All the failing tests have been manually reviewed and the expected outputs updated.
- Regression tests from #537 show a minor change where function application is now expressed as an image rather than a relation projection. A step that was formerly performed in type checking post parsing.

Related issues:
Fixes #521
Fixes #554

### CODE REVIEW

**Comment by Maurice-Byrne** [comment]
> Not the most important thing, but if you change line 6 in the Makefile to GHC-9.2 the build in the CI might progress a bit further.
> 

Looks like the env var is explicitly set in the CI

**Comment by ozgurakgun** [comment]
Oops, yes. Feel free to change it [there](https://github.com/conjure-cp/conjure/blob/master/.github/workflows/build.yml#L27) if you like.

File: etc/hs-deps/stack-9.0.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+resolver: lts-19.24
+packages:
+- '.'
+system-ghc: true
+install-ghc: true
+extra-deps:
+- megaparsec-9.3.0
```
Comment (ozgurakgun) [inline etc/hs-deps/stack-9.0.yaml:7]: Can we remove any of these?

File: etc/hs-deps/stack-9.2.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+resolver: lts-20.2
+packages:
+- '.'
+system-ghc: true
+install-ghc: true
+extra-deps:
```
Comment (ozgurakgun) [inline etc/hs-deps/stack-9.2.yaml:6]: Can we remove any of these?

File: src/test/Conjure/ParserFuzz.hs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import Conjure.Language.AST.Reformer (Flattenable(flatten))
+import Data.Algorithm.Diff (getDiff, getGroupedDiff)
+import Data.Algorithm.DiffOutput (ppDiff)
+import GHC.IO.Handle.FD (stderr)
+
+
+tests :: IO TestTree
+tests = do
+    let baseDir = "tests"
+    allFiles <- readFileIfExists "tests/allfiles.txt"
```
Comment (ozgurakgun) [inline src/test/Conjure/ParserFuzz.hs:30]: make this dynamic using a shell call.

File: etc/hs-deps/stack-9.0.yaml (hunk 1 above)
Comment (Maurice-Byrne) [reply etc/hs-deps/stack-9.0.yaml]: Removed all except megaparsec - 9.3.0 adds better debugging so not essential, but nice to have

Comment (ozgurakgun) [reply etc/hs-deps/stack-9.0.yaml]: perfect. just to be sure: this is a newer version of megaparsec than what's in the lts, right?

Comment (Maurice-Byrne) [reply etc/hs-deps/stack-9.0.yaml]: As far as I can tell it's 9.2.2 in the LTS

File: etc/hs-deps/stack-9.2.yaml (hunk 2 above)
Comment (Maurice-Byrne) [reply etc/hs-deps/stack-9.2.yaml:6]: These all seem to be necessary.

File: src/Conjure/Language/Definition.hs [hunk 4]
```diff
@@ -87,22 +90,21 @@ instance SimpleJSON Model where
     toSimpleJSON m = do
         inners <- mapM toSimpleJSON (mStatements m)
         let (innersAsMaps, rest) = unzip [ case i of JSON.Object mp -> ([mp], []); _ -> ([], [i]) | i <- inners ]
-                                    |> (\ (xs, ys) -> (M.unions (concat xs), concat ys))
+                                    |> (\ (xs, ys) -> ((mconcat) <$> xs, concat ys))
```
Comment (ozgurakgun) [inline src/Conjure/Language/Definition.hs:93]: remove ()

File: src/Conjure/Language/Definition.hs [hunk 5]
```diff
@@ -118,7 +120,7 @@ fromSimpleJSONModel essence json =
                         -- traceM $ show $ "value   " <+> pretty value
                         return $ Just $ Declaration (Letting (Name name) value)
                     _ -> do
-                        logWarn $ "Ignoring" <+> pretty name <+> "from the JSON file."
+                        -- logWarn $ "Ignoring" <+> pretty name <+> "from the JSON file."
```
Comment (ozgurakgun) [inline src/Conjure/Language/Definition.hs:123]: Let's see if we can keep MonadLog here, by adding MonadLog to the caller.

File: src/Conjure/Language/Expression/Op/Internal/Common.hs [hunk 6]
```diff
@@ -236,6 +238,15 @@ operators =
     , ( BinaryOp L_TildeGeq     FNone   ,  400 )
     , ( UnaryPrefix L_Minus           , 2000 )
     , ( UnaryPrefix L_ExclamationMark , 2000 )
+    , ( UnaryPrefix L_Minus           , 2000 )
```
Comment (ozgurakgun) [inline src/Conjure/Language/Expression/Op/Internal/Common.hs:241]: duplicates

File: src/Conjure/Language/Expression/Op/Internal/Common.hs [hunk 7]
```diff
@@ -236,6 +238,15 @@ operators =
...
+    , ( UnaryPrefix L_ExclamationMark , 2000 )
+    ]
+--Functional operators that clash with other constructs so require no spaces
```
Comment (ozgurakgun) [inline src/Conjure/Language/Expression/Op/Internal/Common.hs:244]: Note: this means we are requiring that there will be no space between operator and the list of arguments.

File: src/Conjure/Language/Expression/Op/Negate.hs [hunk 8]
```diff
@@ -20,11 +21,10 @@ instance FromJSON  x => FromJSON  (OpNegate x) where parseJSON = genericParseJSO
 
 instance (TypeOf x, Pretty x) => TypeOf (OpNegate x) where
     typeOf p@(OpNegate a) = do
-        TypeInt t <- typeOf a
-        case t of
-            TagInt -> return ()
+        typeOfA <- typeOf a --TODO: unsure of this refactor
```
Comment (ozgurakgun) [inline src/Conjure/Language/Expression/Op/Negate.hs:24]: remoe this refactor

File: src/Conjure/Language/NewLexer.hs [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            showTrivia :: Trivia -> L.Text
+            showTrivia x = case x of
+              WhiteSpace txt    -> L.fromStrict txt
+              LineComment txt   -> L.fromStrict txt
+              BlockComment txt  -> L.fromStrict txt
+
+reformList :: (Traversable t ,Reformable a) => t a -> L.Text
+reformList =  L.concat  . map reform . toList
+
+emojis :: [Char]
```
Comment (ozgurakgun) [inline src/Conjure/Language/NewLexer.hs:61]: Oz to check why emojis don't work. Add the following as a test case

https://conjure.readthedocs.io/en/latest/tutorials.html#bibd

File: src/Conjure/Language/NewLexer.hs [hunk 10]
```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Conjure.Language.NewLexer (
```
Comment (ozgurakgun) [inline src/Conjure/Language/NewLexer.hs:7]: Remove the old lexer, rename this to Lexer.

File: src/Conjure/Language/Parser.hs [hunk 11]
```diff
@@ -2,1087 +2,1161 @@
 
 module Conjure.Language.Parser
     ( runLexerAndParser
+    , lexAndParse
```
Comment (ozgurakgun) [inline src/Conjure/Language/Parser.hs:5]: Remove commented out code from this file. There was some in the "functionals" file as well.

File: src/Conjure/Language/ParserC.hs [hunk 12]
```diff
@@ -1,698 +1,20 @@
 {-# LANGUAGE RecordWildCards #-}
 
-module Conjure.Language.ParserC ( parseModel ) where
+module Conjure.Language.ParserC (
```
Comment (ozgurakgun) [inline src/Conjure/Language/ParserC.hs:3]: Note: this means we will be allowing operators in the param/solution files, which might make parsing a bit slower for them.

File: src/Conjure/Language/Validator.hs [hunk 13]
```diff
@@ -0,0 +1,2771 @@
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Conjure.Language.Validator where
```
Comment (ozgurakgun) [inline src/Conjure/Language/Validator.hs:5]: Write a short comment describing what this module does. Which is: mapping from AST to Model, capturing all the errors in the process.

File: src/Conjure/UI/TypeScript.hs [hunk 14]
```diff
@@ -1,153 +1,308 @@
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE KindSignatures #-}
 {-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE InstanceSigs #-}
```
Comment (ozgurakgun) [inline src/Conjure/UI/TypeScript.hs:4]: Remove conjure tsdef completely. And this module.

Comment (ozgurakgun) [reply src/Conjure/UI/TypeScript.hs:4]: We made a separate issue for this now.

File: .github/workflows/build.yml [hunk 15]
```diff
@@ -26,7 +26,7 @@ jobs:
 
     strategy:
       matrix:
-        GHC_VERSION: [8.6]
+        GHC_VERSION: ["9.0"]
```
Comment (ozgurakgun) [inline .github/workflows/build.yml:29]: Change the default to 9.2

**Review by ozgurakgun — CHANGES_REQUESTED** [review]

File: src/Conjure/UI/TypeScript.hs (hunk 14 above)
Comment (Maurice-Byrne) [reply src/Conjure/UI/TypeScript.hs]: It broke again so I've removed it now

File: src/Conjure/Language/ParserC.hs (hunk 12 above)
Comment (Maurice-Byrne) [reply src/Conjure/Language/ParserC.hs:3]: New parser (probably mostly from new megaparsec) seems to be at least 25% faster so hopefully this is offset somewhat

File: src/Conjure/Language/NewLexer.hs (hunk 9 above)
Comment (Maurice-Byrne) [reply src/Conjure/Language/NewLexer.hs]: This should now be fixed

**Pull request merged**